EKF/EKFGSF_yaw.cpp

```diff
@@ -0,0 +1,668 @@
+	_ekf_gsf[model_index].P[0][0] = fmaxf(P00-P20*t6+dvxVar*t14+dvyVar*t13-t6*t7 , min_var);
+	_ekf_gsf[model_index].P[0][1] = P01+t12-P21*t6+t7*t10-dvyVar*t2*t3;
+	_ekf_gsf[model_index].P[0][2] = t7;
+	_ekf_gsf[model_index].P[1][0] = P10+t12+P20*t10-t6*t16-dvyVar*t2*t3;
+	_ekf_gsf[model_index].P[1][1] = fmaxf(P11+P21*t10+dvxVar*t13+dvyVar*t14+t10*t16 , min_var);
+	_ekf_gsf[model_index].P[1][2] = t16;
+	_ekf_gsf[model_index].P[2][0] = P20-t8;
+	_ekf_gsf[model_index].P[2][1] = P21+t15;
+	_ekf_gsf[model_index].P[2][2] = fmaxf(P22+dazVar , min_var);
+
+	// force symmetry
+	makeCovSymEKF(model_index);
+}
+
+// Update EKF states and covariance for specified model index using velocity measurement
+void EKFGSF_yaw::updateEKF(const uint8_t model_index)
+{
+	// set observation variance from accuracy estimate supplied by GPS and apply a sanity check minimum
+	float velObsVar = fmaxf(_vel_accuracy, 0.5f);
+	velObsVar *= velObsVar;
+
+	// calculate velocity observation innovations
+	_ekf_gsf[model_index].innov[0] = _ekf_gsf[model_index].X[0] - _vel_NE(0);
+	_ekf_gsf[model_index].innov[1] = _ekf_gsf[model_index].X[1] - _vel_NE(1);
+
+	// copy covariance matrix to temporary variables
+	const float P00 = _ekf_gsf[model_index].P[0][0];
+	const float P01 = _ekf_gsf[model_index].P[0][1];
+	const float P02 = _ekf_gsf[model_index].P[0][2];
+	const float P10 = _ekf_gsf[model_index].P[1][0];
+	const float P11 = _ekf_gsf[model_index].P[1][1];
+	const float P12 = _ekf_gsf[model_index].P[1][2];
+	const float P20 = _ekf_gsf[model_index].P[2][0];
+	const float P21 = _ekf_gsf[model_index].P[2][1];
+	const float P22 = _ekf_gsf[model_index].P[2][2];
+
+	// calculate innovation variance
+	_ekf_gsf[model_index].S[0][0] = P00 + velObsVar;
+	_ekf_gsf[model_index].S[1][1] = P11 + velObsVar;
+	_ekf_gsf[model_index].S[0][1] = P01;
+	_ekf_gsf[model_index].S[1][0] = P10;
+
+	// Perform a chi-square innovation consistency test and calculate a compression scale factor that limits the magnitude of innovations to 5-sigma
+	float S_det_inv = (_ekf_gsf[model_index].S[0][0]*_ekf_gsf[model_index].S[1][1] - _ekf_gsf[model_index].S[0][1]*_ekf_gsf[model_index].S[1][0]);
+	float innov_comp_scale_factor = 1.0f;
+	if (fabsf(S_det_inv) > 1E-6f) {
+		// Calculate elements for innovation covariance inverse matrix assuming symmetry
+		S_det_inv = 1.0f / S_det_inv;
+		const float S_inv_NN = _ekf_gsf[model_index].S[1][1] * S_det_inv;
+		const float S_inv_EE = _ekf_gsf[model_index].S[0][0] * S_det_inv;
+		const float S_inv_NE = _ekf_gsf[model_index].S[0][1] * S_det_inv;
+
+		// The following expression was derived symbolically from test ratio = transpose(innovation) * inverse(innovation variance) * innovation = [1x2] * [2,2] * [2,1] = [1,1]
+		const float test_ratio = _ekf_gsf[model_index].innov[0]*(_ekf_gsf[model_index].innov[0]*S_inv_NN + _ekf_gsf[model_index].innov[1]*S_inv_NE) + _ekf_gsf[model_index].innov[1]*(_ekf_gsf[model_index].innov[0]*S_inv_NE + _ekf_gsf[model_index].innov[1]*S_inv_EE);
```

Comment:
I've added the determinant inverse and matrix inverse to the ekf_gsf struct so that we only calculate them once and have added a helper function to compute the inverse. It adds 5 floats to stack for each filter bank but is worth it in terms of eliminating the duplicated matrix inverse calculation and making the code cleaner.